### ISSUE TITLE
needless_range_loop makes code longer and less understandable

### ISSUE THREAD

**Issue by ethercrow** [rust-lang/rust-clippy#10643]
### Summary

needless_range_loop makes code longer and less understandable

### Lint Name

needless_range_loop

### Reproducer

I tried this code:

```rust
    for i in 0..mem.len() { mem[i] = i; }
```

I saw this happen:

```
warning: the loop variable `i` is used to index `mem`
   --> /src/lib.rs:671:14
    |
671 |     for i in 0..mem.len() { mem[i] = i; }
    |              ^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_range_loop
help: consider using an iterator
    |
671 |     for (i, <item>) in mem.iter_mut().enumerate() { mem[i] = i; }
    |         ~~~~~~~~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~

```

I expected to see this happen: no lints here.

### Version

```text
rustc 1.68.1 (8460ca823 2023-03-20)
binary: rustc
commit-hash: 8460ca823e8367a30dda430efda790588b8c84d3
commit-date: 2023-03-20
host: x86_64-apple-darwin
release: 1.68.1
LLVM version: 15.0.6
```
